File: Coursera/Algorithms_Part_2/Ex31ThreeWayRadixSort.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.StdOut;
import edu.princeton.cs.algs4.StdRandom;
import edu.princeton.cs.algs4.Stopwatch;

public class Ex31ThreeWayRadixSort {

    public static void sort(String[] s) {
        sort(s, 0, s.length - 1, 0);
    }

    private static void sort(String[] s, int lo, int hi, int d) {
        if (hi <= lo) return;
        int lt = lo;
        int gt = hi;
        int v = charAt(s[lo], d);
        int k = lo;
        while (k <= gt) {
            if (charAt(s[k], d) < v) exch(s, k++, lt++);
            else if (charAt(s[k], d) > v) exch(s, k, gt--);
            else k++;
        }

        sort(s, lo, lt - 1, d);
        if (v >= 0) sort(s, lt, gt, d + 1);
        sort(s, gt + 1, hi, d);
    }

    private static int charAt(String s, int d) {
        if (d < s.length()) return s.charAt(d);
        return -1;
    }

    private static boolean less(String a, String b) {
        return a.compareTo(b) < 0;
    }

    private static void exch(String[] s, int i, int j) {
        String temp = s[i];
        s[i] = s[j];
        s[j] = temp;
    }

    private static boolean isSorted(String[] s) {
        for (int i = 1; i < s.length; i++)
            if (less(s[i], s[i - 1])) return false;
        return true;
    }

    public static void main(String[] args) {
        int n = StdRandom.uniform(5, 1000000);
        String[] test = new String[n];
        StdOut.println("Original:");
        int maxLength = 50;
        for (int i = 0; i < n; i++) {

            int len = StdRandom.uniform(3, 20);
            StringBuilder builder = new StringBuilder();
            if (i > 1 && StdRandom.bernoulli()) {
                int ri = StdRandom.uniform(i);
                builder.append(test[ri]);
                len = Math.min(maxLength, StdRandom.uniform(5) + test[ri].length());
            }

            for (int j = 0; j < len; j++) {
                if ((j > 0 || i > 0) && StdRandom.bernoulli()) {
                    if (j > 0) builder.append(builder.charAt(j - 1));
                    else builder.append(test[i - 1].charAt(0));
                }
                else {
                    int ci = StdRandom.uniform('a', 'z' + 1);
                    builder.append((char) ci);
                }
            }
            test[i] = builder.toString();
            StdOut.println(builder.toString());
        }

        Stopwatch stopwatch = new Stopwatch();
        sort(test);
        double timeElapsed = stopwatch.elapsedTime();

        StdOut.println();
        StdOut.println("Sorted:");
        for (String s : test)
            StdOut.println(s);

        StdOut.println();
        StdOut.println("n:" + n + " elapsed time : " + timeElapsed);
        StdOut.println("isSorted? " + isSorted(test));
    }
}

File: PythonLearnig/Algorithms_Part_2/Ex31ThreeWayRadixSort.py
def sort(unsorted_array):

    def char_at(s, d):
        if len(s) <= d:
            return -1
        return ord(s[d])

    def exch(i, j):
        temp = unsorted_array[i]
        unsorted_array[i] = unsorted_array[j]
        unsorted_array[j] = temp

    def sort_i(lo, hi, d):
        if hi <= lo:
            return

        lt = lo
        gt = hi
        k = lo
        v = char_at(unsorted_array[lo], d)
        while k <= gt:
            c = char_at(unsorted_array[k], d)
            if c < v:
                exch(k, lt)
                k += 1
                lt += 1
            elif c > v:
                exch(k, gt)
                gt -= 1
            else:
                k += 1

        # print(f'sort_i lo: {lo} hi: {hi} lt: {lt} gt: {gt} d: {d}')
        sort_i(lo, lt - 1, d)
        if v > 0:
            sort_i(lt, gt, d + 1)
        sort_i(gt + 1, hi, d)

    sort_i(0, len(unsorted_array) - 1, 0)


def test_three_way_radix_sort():
    import random

    unsorted_array = []
    for i in range(15):
        unsorted_array.append('')
        for j in range(random.randint(5, 10)):
            unsorted_array[i] += chr(random.randint(ord('A'), ord('B')))

    for _ in range(15):
        unsorted_array.insert(random.randint(0, len(unsorted_array) - 1),
                              unsorted_array[random.randint(0, len(unsorted_array) - 1)])

    print()
    print('Original:')
    for key in unsorted_array:
        print(f'\t{key}')
    sort(unsorted_array)
    print('Sorted  :')
    for key in unsorted_array:
        print(f'\t{key}')

    print()
    is_sorted = True
    for i in range(14):
        if unsorted_array[i] > unsorted_array[i + 1]:
            print(f'{unsorted_array[i]} > {unsorted_array[i + 1]}')
            is_sorted = False
    print(f'Is Sorted? {is_sorted}')


if __name__ == '__main__':
    test_three_way_radix_sort()
